Feat: signup, login, and getting user public info

dexterklui opened the pull request.

dexterklui committed:
feat: add User entity

dexterklui committed:
feat: add UserRepo repository

dexterklui committed:
feat: add Role entity to define user privileges

dexterklui committed:
feat: add RoleRepo repository class

dexterklui committed:
feat(Role): add parametrised constructor

dexterklui committed:
feat: add DataLoader component to create Role

dexterklui committed:
feat(User): add account auth related fields

dexterklui committed:
feat: add AuthUser class for user authentication

It implements UserDetails and has a User

dexterklui committed:
feat(RoleRepo): add findByRole method

dexterklui committed:
feat(UserRepo): add findByUsername method

dexterklui committed:
feat: add RoleService service class

dexterklui committed:
feat: add auth related annotations in main class

dexterklui committed:
feat: add AuthUserService service class

dexterklui committed:
refactor: remove unused import

dexterklui committed:
feat: add RsaKeyProperties record

dexterklui committed:
build: add "OAuth2 Resource Server" dependency

dexterklui committed:
refactor: extract createUser method to UserService

dexterklui committed:
refactor: move AuthUserService to security package

dexterklui committed:
feat: add sample public & private key

For convenience of development

dexterklui committed:
config: specify location of public & private key

dexterklui committed:
feat: add annotation for RSA Key properties

dexterklui committed:
feat: add SecurityConfig

- Create Beans for authentication
- Config filter chain to require authentication for requests other than
  user registration
- Allow CORS to domain configured in application properties

dexterklui committed:
feat: add TokenService for authentication

dexterklui committed:
refactor: remove print and super() statements

dexterklui committed:
feat(User): add no-args constructor and validation

dexterklui committed:
feat: add InternalServerError custom exception

For 500 response

dexterklui committed:
feat(UserService): create a new user

dexterklui committed:
feat: create USER role if not exists at startup

dexterklui committed:
feat: add API endpoint to handle user registration

dexterklui committed:
feat: add constraint to Role entity

dexterklui committed:
feat: conflict response for user registration

dexterklui committed:
refactor: rename to InternalServerErrorException

dexterklui committed:
feat: add handler for InternalServerErrorException

Also minor change to the message body

dexterklui committed:
feat: 400 response for response failing @Valid check

dexterklui committed:
docs: add JavaDoc for GlobalExceptionHandler

dexterklui committed:
feat: better response for requests failing @Valid check

dexterklui committed:
refactor: extract value to variables

dexterklui committed:
feat: add RecordNotFoundException for 404 response

dexterklui committed:
feat: API endpoint to get user public info by id

dexterklui committed:
feat: API endpoint for user login

dexterklui committed:
refactor(User): rename foreign key column in DB

essteer commented on src/main/java/com/fdmgroup/forex/exceptionhandlers/GlobalExceptionHandler.java:
NB presumably we should add the `InsufficientFundsException` to this handler once the portfolio feature branch is merged.

dexterklui replied:
Exactly

dexterklui committed:
fix: misleading 400 response message

essteer commented on src/main/java/com/fdmgroup/forex/models/User.java:
```diff
@@ -0,0 +1,152 @@
+	private UUID id;
+
+	@NotNull(message = "'username' must be present.")
+	@NotBlank(message = "'username' cannot be blank, i.e. is empty or contains only white spaces.")
```
Consider rewording to:
"'username' cannot be blank or contain only whitespace characters."

dexterklui replied:
Commit 5b60182 solved the issue.

essteer commented on src/main/java/com/fdmgroup/forex/models/User.java:
```diff
@@ -0,0 +1,152 @@
+	private String username;
+
+	@NotNull(message = "'email' must be present.")
+	@Email(message = "'email' must be a valid Email address.")
```
'email' rather than 'Email' mid-sentence

dexterklui replied:
Commit 5b60182 solved the issue.

essteer commented on src/main/java/com/fdmgroup/forex/models/User.java:
```diff
@@ -0,0 +1,152 @@
+	private String email;
+
+	@NotNull(message = "'password' must be present.")
+	@NotBlank(message = "'password' cannot be blank, i.e. is empty or contains only white spaces.")
```
Consider rewording to:
"'username' cannot be blank or contain only whitespace characters."

dexterklui replied:
Commit 5b60182 solved the issue

essteer commented on src/main/resources/certs/private.pem:
@dexterklui did you intend to commit this?

dexterklui replied:
Include them for the convenience of pulling and running locally for development and testing
Updated README to remind users to generate their own key pairs

essteer reviewed (commented):
Nice work - the Spring Security side is unfamiliar to me, let's review together with Marco to see the functionality if that's possible at this stage.

dexterklui committed:
docs: update README about generating RSA key pair

hksoaa commented on src/main/java/com/fdmgroup/forex/controllers/UserController.java:
```diff
@@ -0,0 +1,54 @@
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
```
annotation.*

dexterklui replied:
The commit c8e05a9 solved the issue

dexterklui committed:
feat: improve error response message

dexterklui committed:
refactor: use '*' to reduce import statements

dexterklui committed:
refactor: move exception throwing to service layer

dexterklui committed:
Merge branch 'develop' into feat/user-login-signup

dexterklui committed:
test: fix: use the new User constructor

dexterklui committed:
feat: add handler for InsufficientFundException

dexterklui committed:
fix(portfolio): add list initialisation

dexterklui committed:
fix(User): forgot to set a field in constructor

dexterklui committed:
refactor: make foreign key column name consistent

dexterklui committed:
feat(Portfolio): add overloaded constructor

For creating a new portfolio with a newly created user

dexterklui committed:
feat: create a portfolio for newly created user

dexterklui committed:
refactor(UserService): extract to method

dexterklui committed:
style(application.properties): categorise settings

dexterklui committed:
feat: allow setting default role with .env

essteer commented on src/main/java/com/fdmgroup/forex/models/Portfolio.java:
```diff
@@ -25,10 +25,18 @@ public class Portfolio {
 
-    public Portfolio() {}
+    public Portfolio() {
+        assets = List.of();
```
Good idea.

essteer commented on src/test/java/com/fdmgroup/forex/models/PortfolioAssetTest.java:
NB This class is renamed to `AssetTest` since the `PortfolioAsset` class is now `Asset`.

dexterklui replied:
resolved by reverting and merging with #7

essteer commented on src/test/java/com/fdmgroup/forex/models/PortfolioAssetTest.java:
The indentation seems excessive  (now 8 spaces instead of 4)

dexterklui replied:
As discussed, leave the formatting style for now

essteer commented on src/test/java/com/fdmgroup/forex/models/PortfolioTest.java:
As per `PortfolioAssetTest.java` / `AssetTest.java` - excessive indentation?

dexterklui replied:
As discussed, leave the formatting style for now

dexterklui committed:
feat: only expose user public info in portfolio

In a response body including the portfolio object

dexterklui committed:
feat!: change get Portfolio API

Remove API to get Portfolio by user object, because GET request
shouldn't have a request body. And the client shouldn't sending an
entire User object including matching hashed password anyway.

Instead add API to get Portfolio by username.

dexterklui committed:
feat: by default ignore user sensitive fields

And remove UserPublicInfoDTO and GetPortfolioDTO because now by default
already ignore sensitive fields from User object, using @JsonIgnore

dexterklui committed:
feat!: add API to find portfolio by user id

BREAKING CHANGE: Remove find portfolio by username. Because this can
change.

dexterklui committed:
feat!: add API to find FX rate by currency code

BREAKING CHANGE: remove API to find FX rate by currency object. Because
a GET request shouldn't have a request body.

essteer commented on src/main/java/com/fdmgroup/forex/services/FxRateService.java:
See separate pull request - method added to FxRateRepo to implement this change, without needing to add CurrencyRepo dependency to FxRateService

essteer replied:
#7

dexterklui replied:
resolved by reverting and merging with #7

essteer commented on src/main/java/com/fdmgroup/forex/controllers/FxRateController.java:
See #7

dexterklui replied:
resolved by reverting and merging with #7

essteer commented on src/main/java/com/fdmgroup/forex/controllers/PortfolioController.java:
See #7

dexterklui replied:
resolved by reverting and merging with #7

essteer commented on src/main/java/com/fdmgroup/forex/services/PortfolioService.java:
See #7

dexterklui replied:
resolved by reverting and merging with #7

dexterklui committed:
Revert "feat!: add API to find FX rate by currency code"

This reverts commit e70b44fc7f8e5178fdc740ccce648e23097985d6.

dexterklui committed:
Merge branch 'develop' into feat/user-login-signup

essteer reviewed (approved):
Looks good to me!

dexterklui committed:
fix: user registration

Adding @JsonIgnore to fields in previous commit 1fc889b broke user
registration because the annotation also prevents deserialization.
